keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1248 @@
+
+It is not possible to be 100% sure of this property, in fact, there may be new validations added to patch bugs in existing ones. But we can mitigate this risk:
+
+0. PREREQUISITE - No field can go thru migration without a robust test for the field in question, which proves that it is validated correctly before the change and after. 
+
+Many existing validation tests do not account for all required valdiation cases.  It is possible that if there is unintentionally a change some small aspect of it (making validation for an existing field weaker or stronger) there could be errors in the migration. The only real solution is to scrutinize and retool tests.
+
+1. Catch and fix any significant differences in schema validations
+
+To help catch these errors faster, we will also refactor all e2e tests to enable this feature and check for validation error differences.
+
+2. Minimize the impact of any new changes introduced
+
+As mentioned in the section - “Ratcheting” `validation-gen` is capable of supporting any desired ratcheting behaviour.  Validation Ratcheting would reduce impact to users by allowing any new schema validations errors due to this feature to be ignored for unchanged portions of an object.
+
+#### Risk: Added latency to API request handling.
+
+**NOTE:** The validation path has NEVER been a target for efficiency optimizations, it is likely there is currently low-hanging fruit with respect to optimizing.  Currently the code does such things as deep-copies, use reflect, build maps of 3 items, etc.  
+
+The design decision to declaratively validate API versions has performance implications.
+
+Today, the apiserver's request handler decodes incoming versioned API requests to the versioned type and then immediately performs an "unsafe" conversion to the internal type when possible. All subsequent request processing uses the internal type as the "hub" type. Validation is written for the internal type and so are per-resource strategies.
+
+Assuming we use the recommended plan of declarative validation operating on versioned types, the internal type will no longer be responsible for validation.
+
+If we were to convert from the internal back to the version of the API request for validation, we would introduce one additional conversion. If we make this an "unsafe" conversion, then it will be low cost for the vast majority of requests.
+
+We will benchmark this approach and plan to use it for beta.
+
+Long term, we could do better:
+
+Since the internal type will no longer be used for validation, it becomes a lot less important. It is still important to have a hub type. But why not pick one of the versioned types to be the hub type? The vast majority of APIs only have one version anyway. The obvious candidate version to choose for the hub version would be the preferred storage version.
+
+Switching to a versioned type for the hub type would have a few implications:
```

Comment:
also requires rewriting in tree admission plugins.

Reply:
Updated the implications to state switching to a versioned type for the hub type would also require rewriting in tree admission plugins.

> *   Would require rewriting in tree admission plugins.